### DISCUSSION TITLE
Vivo desk compatability

### DISCUSSION THREAD

**Discussion by ajgrinnell** [tjhorner/wifi-desk-controller#2] [Reverse Engineering Help]
I have a Vivo desk (https://www.amazon.com/gp/product/B07V6ZTYZ7/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) that uses the LoctekMotion HS01B-1 board. This is not currently compatible and I'm looking for a pinout that will work. Open to any suggestions.

Here is the pinout of the board:

RJ45 pin | Name | Original Cable Color | Ethernet cable color (T568B)
-- | -- | -- | --
8 | +5V (VDD) | Yellow | Brown
7 | GND | Blue | White-Brown
6 | TX | Black | Green
5 | RX | Green | White-Blue
4 | PIN 20 | Red | Blue
3 | (unknown) | Purple | White-Green
2 | SWIM | White | Orange
1 | RES | Brown | White-Orange

+5v, gnd, tx, rx seem pretty straight forward, the other 4 pins need translating.

**Comment by tjhorner** [comment]
Hmm, what do you mean a pinout that works? As it is now, +5V and GND are in the wrong place so the board shouldn't be compatible at all, lol. Are you using an RJ45 breakout to remap the pins or something like that?

Anyway, I did some light Googling and it looks like others have already reverse engineered the LoctekMotion board you have: https://github.com/iMicknl/LoctekMotion_IoT

It looks like the keypad communicates everything over UART, including the presets, up/down commands, etc. [Here's an example ESPHome config](https://github.com/iMicknl/LoctekMotion_IoT/blob/5d9f9c01aa951d8ca24eac3198709b755d330e1b/packages/esphome/flexispot_e5b.yaml#L73-L122) I found from the repo. Seems simple enough. Could you try sending a few of those serial messages (i.e., `0x9b, 0x06, 0x02, 0x02, 0x00, 0x0c, 0xa0, 0x9d` for the down command)? That should take care of the communication _to_ the desk.

As for reading the height, there also seems to be an [ESPHome component](https://github.com/iMicknl/LoctekMotion_IoT/blob/main/packages/esphome/desk_height_sensor.h) for that. It looks very similar to the protocols that Uplift/Jarvis use, so you might be able to use [my external component](https://github.com/tjhorner/esphome-standing-desk#external-component) for that. If not, it should be easy enough to import the component from that repo.

I don't think the other pins matter, at least for what we are trying to do.

Let me know if any of these work out for you. If they do, I will work with the maintainer of the LoctekMotion_IoT repo to add support for those desks and provide a stock firmware for them, if possible.

For v2 of the board, I'd like to add support for different pinouts (via dip switches or something), but I will look into making adapters of some sort for the v1 board, so that you can use desks that have different pinouts without too much hackery.

**Reply by ajgrinnell** [reply]
So really I would only need +5v, gnd, tx, and rx- the other pins dont seem to be needed then from how I am understanding this. if thats the case Ill just make a couple cables to get the pins in the right place.

**Reply by tjhorner** [reply]
Yep, that's correct. The board expects +5V on pin 5, and GND on pin 3. You can wire the RX and TX pins anywhere else.

**Edit**: Actually, it looks like you will need to wire "PIN 20" as well. It's used as a ["Virtual Screen"](https://github.com/iMicknl/LoctekMotion_IoT/blob/5d9f9c01aa951d8ca24eac3198709b755d330e1b/packages/esphome/flexispot_e5b.yaml#L65-L71) in the config. I'm not sure why, but if I had to guess it's probably because the desk needs to detect the presence of a more advanced keypad, rather than a simple up/down one.

**Reply by ajgrinnell** [reply]
Awesome, I'm going to give this a shot tonight. Are the pins of the rj45s bridged together, or is the D1 in the middle, ie if the D1 is removed, will communication from keypad to controller stop?

**Reply by tjhorner** [reply]
Yes, they are bridged together. Removing the D1 Mini won't stop communication between the two jacks

**Reply by ajgrinnell** [reply]
Getting closer, but no luck yet. the pins seem to be ok now, but only if the D1 is removed. I can move the desk until the D1 comes online, then the keypad goes blank. I thought there may be a short somewhere, but so far nothing is proving that theory. Open to any ideas.

**Reply by ajgrinnell** [reply]
Problem appears to be between the keypad and the D1 Mini. If I unplug the controller, I can control the desk via Homeassistant. Checking the pinout of the cable facing the keypad again.... Even tried swapping Tx/Rx, no luck yet but so very close.

**Reply by ajgrinnell** [reply]
Everything works as long as I have the Tx disconnected from the keypad. Can still control from HA, and now the changes are displayed on the keypad, but still cannot control from the keypad.

**Reply by iMicknl** [reply]
@ajgrinnell feel free to join our Discord group for Loctekmotion hacking; https://discord.gg/dvepqK79. There are quite some people with experience, that perhaps can help you debug.

**Reply by Jolanrensen** [reply]
@ajgrinnell any luck so far? Got a HS01B-1 desk too, plus an upsy desky, some ethernet connectors and a crimping tool, so maybe I can try something myself too :)

**Reply by mgomez009** [reply]
@Jolanrensen any luck with this?

**Reply by Jolanrensen** [reply]
@mgomez009 Well yes, I've done it with an upsy-desky https://github.com/Jolanrensen/LoctekMotion_IoT/blob/main/packages/esphome/hs01b-1_upsy_desky.yaml

You can read up on what I did here https://github.com/tjhorner/upsy-desky/issues/13 and here https://github.com/tjhorner/upsy-desky/issues/5. It's not perfect and required a lot of work and tinkering, but it still works today :)

**Reply by mgomez009** [reply]
Thank you! So I have now created the adapter cable but I have not added the female to male adapter instead I used a breakout board. It seems to be giving me trouble with the code though. I am not sure where it is going wrong. I have an e5b but the code that I copied over seems like it is not working with my desk

https://github.com/tjhorner/wifi-desk-controller/assets/109258015/d583c86b-82ae-43b8-a0e9-3dcd57d0e0b5
